test/test_autograd.py

```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
         mean_combined = torch.stack(feat_combined).mean()
         mean_combined.backward()
 
+    def test_autograd_views_codegen(self):
+        # This is not necessarily the absolute correct behavior, but this is the current
+        # one. This test is here to make sure that any change to this behavior is detected
+        # and not silent. The TODOs below mark the places with unexpected behavior.
+        # Note that any change in these test will be BC-breaking and should be done carefully.
+
+        # This test checks the behavior of two codegen functions (view_as and unbind)
+        # with respect to view tracking and inplace operation on the output.
+
+        def run_test(grad_mode, requires_grad, is_view, should_raise_tuple):
+            def get_view_as():
+                inp = torch.rand(2, requires_grad=requires_grad).clone()
+                with torch.set_grad_enabled(grad_mode):
+                    return inp.view_as(inp)
+
+
+            def get_unbind():
+                inp = torch.rand(2, requires_grad=requires_grad).clone()
+                with torch.set_grad_enabled(grad_mode):
+                    return inp.unbind()
+
+            def maybe_check_raise(fn, should_raise):
+                self.assertTrue(should_raise is None or isinstance(should_raise, str))
+                if should_raise is not None:
+                    with self.assertRaisesRegex(RuntimeError, should_raise):
+                        fn()
+                else:
+                    fn()
+
+            # Are they differentiable views?
+            out = get_view_as()
+            self.assertTrue(out._is_view() == is_view)
+
+            out = get_unbind()
+            self.assertTrue(out[0]._is_view() == is_view)
+            self.assertTrue(out[1]._is_view() == is_view)
+
+            # Are inplace allowed?
+            out = get_view_as()
+            maybe_check_raise(lambda: out.add_(1), should_raise_tuple[0])
+
+            out = get_unbind()
+            maybe_check_raise(lambda: out[0].add_(1), should_raise_tuple[1])
+            maybe_check_raise(lambda: out[1].add_(1), should_raise_tuple[2])
+
+
+        # should_raise contains None if it should not raise
+        # should_raise contains a string of the error if it should raise
+        # The 3 elements are for view_as, first output of unbind and second output of unbind
+        run_test(grad_mode=True, requires_grad=False, is_view=True,
+                 should_raise_tuple=(None, None, None))
+        # TODO: Second should_raise should not be None below
+        run_test(grad_mode=True, requires_grad=True, is_view=True,
+                 should_raise_tuple=(None, None, "diff_view_meta->output_nr_ == 0 INTERNAL ASSERT FAILED"))
```

Comment:
The contract of INTERNAL ASSERT is that it should be impossible to trigger it from userland (aka Python). So this is just a straight up bug; either the assert is incorrectly classified, or we're doing something wrong here

Reply:
Yes, both the 2nd and 3rd here should raise an error stating that we're doing an inplace on the output of a function where it's not allowed.